### PR TITLE
Stations

### PR DESCRIPTION
**Pull request by revol-xut**
# General

This pull request implements more endpoints of #2 

## Endpoints Implemented

- **GET /stations** (list)
- **POST /stations** (create)
- **GET /stations/id** (info)
- **PUT /stations/id** (update)
- **DELETE /stations/id** (delete)
- **POST /stations/id/approve** (approving stations)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -95,11 +95,14 @@ async fn main() -> std::io::Result<()> {
...
             .route("/region/{id}", web::put().to(routes::region::region_update))
             .route("/region/{id}", web::get().to(routes::region::region_info))
-            .route(
-                "/region/{id}",
-                web::delete().to(routes::region::region_delete),
-            )
-    })
+            .route("/region/{id}", web::delete().to(routes::region::region_delete))
+            .route("/station", web::post().to(routes::station::station_create))
+            .route("/station", web::get().to(routes::station::station_list))
```
Comment (oxapentane) [inline src/main.rs:100]: maybe worth doing `/station/all`? Also would help to avoid confusion with creation POST

File: src/routes/station.rs [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    pub architecture: Option<i32>,
+    pub device: Option<i32>,
+    pub elevation: Option<f64>,
+    pub antenna: Option<i32>,
+    pub telegram_decoder_version: Option<String>,
+    pub notes: Option<String>
+}
+
+
+/// holds all the necessary information that are required to create a new station
```
Comment (oxapentane) [inline src/routes/station.rs:35]: copypasta

File: src/routes/station.rs [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        id: Uuid::new_v4(),
+        token: Some(random_token),
+        name: request.name.clone(),
+        lat: request.lat,
+        lon: request.lon,
+        region: request.region,
+        owner: user_session.id,
+        approved: false,
+        deactivated: false,
+        public: false,
```
Comment (oxapentane) [inline src/routes/station.rs:133]: Shouldn't also be directly configurable by the request? Or am I brain-farting?

File: src/routes/station.rs [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+        match stations
+            .filter(public.eq(true).or(owner.eq(user_session.id)))
+            .load::<Station>(&mut database_connection)
+        {
+            Ok(all_station) => Ok(web::Json(all_station)),
+            Err(e) => {
+                error!("error while fetching the config {:?}", e);
+                return Err(ServerError::InternalError);
+            }
+        }
```
Comment (oxapentane) [inline src/routes/station.rs:203]: How public we want to keep the API? I guess list of public stations we can also give to non-logged-in ppl. I guess we should also talk about that at some point

File: src/routes/station.rs [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+
+    // get currently logged in user
+    let user_session = fetch_user(identity, &mut database_connection)?;
+
+    use tlms::schema::stations::dsl::stations;
+    use tlms::schema::stations::{id, name, lat, lon, public, architecture, device, elevation, antenna, telegram_decoder_version, notes};
+    
+    let relevant_station = match stations
+        .filter(id.eq(path.0))
+        .limit(1)
```
Comment (oxapentane) [inline src/routes/station.rs:241]: Is this a diesel fucky? If not, we should clearly throw HTTP 5xx if id is non-unique...

File: src/routes/station.rs [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+    // get currently logged in user
+    let user_session = fetch_user(identity, &mut database_connection)?;
+
+    use tlms::schema::stations::dsl::stations;
+    use tlms::schema::stations::{id, deactivated};
+    
+    let relevant_station = match stations
+        .filter(id.eq(path.0))
+        .limit(1)
```
Comment (oxapentane) [inline src/routes/station.rs:312]: same as above

File: src/routes/station.rs [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        .first::<Station>(&mut database_connection)
+    {
+        Ok(possible_station) => possible_station,
+        Err(e) => {
+            error!("error while searching for mentioned station {:?}", e);
+            return Err(ServerError::InternalError);
+        }
+    };
+
+    warn!("trying to deleting station! : {}", relevant_station.id);
```
Comment (oxapentane) [inline src/routes/station.rs:322]:
```suggestion
    warn!("trying to delete station! : {}", relevant_station.id);
```

File: src/routes/user.rs [hunk 8]
```diff
@@ -471,10 +471,10 @@ pub async fn user_info(
     Ok(web::Json(user))
 }
 
-/// Returns information about the currently authenticated user
+/// Returns list of users
 #[utoipa::path(
     get,
-    path = "/user/list",
+    path = "/user",
```
Comment (oxapentane) [inline src/routes/user.rs:477]: same as with stations, maybe better '/user/list'

**Review by oxapentane — COMMENTED** [review]
Please see comments, but in general looks gud

File: src/routes/station.rs (hunk 5 above)
Comment (oxapentane) [reply src/routes/station.rs:241]: And also shit into our logs hard if id is non-unique

File: src/main.rs (hunk 1 above)
Comment (revol-xut) [reply src/main.rs:100]: the problem with this is that this is not really REST anymore.

File: src/routes/user.rs (hunk 8 above)
Comment (revol-xut) [reply src/routes/user.rs:477]: REST Endpoint if u want to discuss about the endpoint schema maybe take a lot at #2

File: src/routes/station.rs (hunk 3 above)
Comment (revol-xut) [reply src/routes/station.rs:133]: valid point

File: src/routes/station.rs [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    // get currently logged in user
+    let user_session = fetch_user(identity, &mut database_connection)?;
+
+    if !user_session.is_admin() {
+        return Err(ServerError::Unauthorized);
+    }
+
+    // if the region doesn't exist we can directly dispose of the request
+    use tlms::schema::regions::dsl::regions;
+    use tlms::schema::regions::id;
```
Comment (MarcelCoding) [inline src/routes/station.rs:102]: Is this a diesel thing, or why do you don't use global imports. (If the names are overlapping, you could use aliases.)

**Review by MarcelCoding — COMMENTED** [review]
Expect that stuff that is already being discussed, this seems to be fine.

File: src/routes/station.rs (hunk 2 above)
Comment (revol-xut) [reply src/routes/station.rs]: done

File: src/routes/station.rs (hunk 4 above)
Comment (revol-xut) [reply src/routes/station.rs]: done

File: src/routes/station.rs (hunk 5 above)
Comment (revol-xut) [reply src/routes/station.rs]: done

File: src/routes/station.rs (hunk 6 above)
Comment (revol-xut) [reply src/routes/station.rs]: done

File: src/routes/station.rs (hunk 9 above)
Comment (revol-xut) [reply src/routes/station.rs:102]: done optimized the imports

**Review by MarcelCoding — APPROVED** [review]

**Review by oxapentane — APPROVED** [review]
THIS IS VERY GOFFFIK AND NOT POSER AT ALL!!!!111

**Pull request merged**